extensions/data-transfer/portability-data-transfer-generic/src/main/java/org/datatransferproject/datatransfer/generic/auth/OAuthTokenManager.java

```diff
@@ -0,0 +1,171 @@
+package org.datatransferproject.datatransfer.generic.auth;
+
+import static java.lang.String.format;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonParseException;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.PropertyNamingStrategies;
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import javax.annotation.Nullable;
+import okhttp3.FormBody;
+import okhttp3.MediaType;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import org.datatransferproject.api.launcher.Monitor;
+import org.datatransferproject.spi.transfer.types.InvalidTokenException;
+import org.datatransferproject.types.transfer.auth.AppCredentials;
+import org.datatransferproject.types.transfer.auth.TokensAndUrlAuthData;
+
+@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
+class RefreshTokenResponse {
+  private final String accessToken;
+  private final Optional<String> refreshToken;
+  private final String tokenType;
+  private final Optional<String> expiresIn;
+
+  @JsonCreator
+  public RefreshTokenResponse(
+      @JsonProperty(value = "access_token", required = true) String accessToken,
+      @Nullable @JsonProperty("refresh_token") String refreshToken,
+      @JsonProperty(value = "token_type", required = true) String tokenType,
+      @Nullable @JsonProperty("expires_in") String expiresIn) {
+    this.accessToken = accessToken;
+    this.refreshToken = Optional.ofNullable(refreshToken);
+    this.tokenType = tokenType;
+    this.expiresIn = Optional.ofNullable(expiresIn);
+  }
+
+  public String getAccessToken() {
+    return accessToken;
+  }
+
+  public Optional<String> getRefreshToken() {
+    return refreshToken;
+  }
+
+  public String getTokenType() {
+    return tokenType;
+  }
+
+  public Optional<String> getExpiresIn() {
+    return expiresIn;
+  }
+}
+
+/**
+ * Utility to manage {@link TokensAndUrlAuthData} containing OAuth refresh and access tokens.<br>
+ * See {@see #withAuthData} for how to use the auth data managed by this class.
+ */
+public class OAuthTokenManager {
```

Comment:
I'm surprised this isn't already part of the DTP framework. Doesn't it have to be reimplemented like this for each extension? Not something that needs to be addressed in this PR for sure, just asking!